### DISCUSSION TITLE
Shelly Pro 4PM display colors swapped

### DISCUSSION THREAD

**Discussion by phiten** [arendst/Tasmota#17971] [Q&A] [answered]
@gemu2015 the config file for the Shelly Pro 4PM is here: https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7735S_Pro4PM_display.ini

But the colors blue and red are swapped. Which values do I need to have a look at? Couldn't find docs for that.

**Comment by gemu2015** [comment, marked as answer]
try this: cmd 36 is RGB,BGR order
36,1,C8  ->  36,1,C0

**Reply by phiten** [reply]
That sadly didn't change anything.

**Reply by gemu2015** [reply]
yes i see, cmd 36 is overwritten in rotation code,  clr bit 3 on all 4 rotations:

like this
R,36
:0,60,00,00,00
:1,C4,1A,01,01
:2,A0,01,1A,02
:3,00,1A,01,03

**Reply by phiten** [reply]
Awesome thanks that worked. Will push the changes in a moment.

**Reply by phiten** [reply]
Hey @gemu2015,

I realized that I may have made a mistake by enabling the dark mode by default here: https://github.com/arendst/Tasmota/commit/da81d25fb5cd5a74e1cefaee31d339171f2b3aa5

Because the LVGL switches are then purple instead of green e.g.

So in my local config I now reverted the change and trying to figure out the colors again. Maybe you could have a look again:
#000000 is white
#0000FF is yellow
#FFFFFF is black
#FF0000 is turquoises
#00FF00 is purple

**Reply by gemu2015** [reply]
internally we work with RGB565
i don't use LVGL  don't know how LVGL translates to RGB565
20/21  is display color invert
and bit 3 of code 36 ist flip BR
so you have 4 permutations.
i would try all and compare with existing LVGL pictures

**Comment by ggrote** [comment]
How did you flash the Shelly Pro 4PM with tasmota? Couldn't find any information on this.

**Reply by Noschvie** [reply]
@ggrote Have you got it flashed ? Which template are you using? Thanks!

**Reply by ggrote** [reply]
No haven't tried it yet.

**Reply by Noschvie** [reply]
[MGOS32 to Tasmota](https://github.com/tasmota/mgos32-to-tasmota32) shall do it, will try it...

**Comment by ramponis** [comment]
Can someone give me some details how to flash Shelly pro 4PM to tasmota?

Thank you

**Reply by Noschvie** [reply]
[mgos32-to-tasmota32](https://github.com/tasmota/mgos32-to-tasmota32) works for me

**Reply by ramponis** [reply]
Do you remember the shelly firmware version that you came from?
And also the mgos32 version that you installed?
I want to avoid any truoble like this:
[issue85](https://github.com/tasmota/mgos32-to-tasmota32/issues/85)

Thanks

**Reply by Noschvie** [reply]
mgos32-to-tasmota32 [v13.4.1](https://github.com/tasmota/mgos32-to-tasmota32/tree/v13.4.1), don't remember the Shelly firmware version, but didn't connect the Shelly to the internet.

**Reply by ramponis** [reply]
Tested but my pro 4pm seems to be blocked.
After flashing with mgos32, it reboots and reload shelly firmware.
Is there any information about the serial interface?
I want to try to flash it with serial

**Reply by sfromis** [reply]
The teardown images here shows a 6-pin socket adjacent to the Ethernet RJ45 connector
https://karlquinsland.com/shelly-pro-4pm-teardown/

That is very probably the "usual" Shelly flashing connector, with pins
> TXD, RXD, 3V3, RST, IO0, GND

Images does not show the orientation, but it should be easy to verify which is GND.

**Reply by ramponis** [reply]
I will try and let you know...
Can someone that have already flashed it with MGOS32 post the template?

Thank you

**Reply by sfromis** [reply]
The Auto-config setup for the device includes the template along with further things also needed.
https://github.com/tasmota/autoconf/tree/main/raw/esp32/Shelly_Pro_4PM
